File: problems/1700-number-of-students-unable-to-eat-lunch/python/1700_number_of_students_unable_to_eat_lunch.py
from typing import List


class Solution:
    def countStudents(self, students: List[int], sandwiches: List[int]) -> int:
        count: int = 0

        while len(students) != 0 and len(sandwiches) != 0:
            student: int = students[0]
            sandwich: int = sandwiches[0]

            before: str = "".join([str(s) for s in students])

            if student != sandwich:
                students.append(students.pop(0))
            else:
                students.pop(0)
                sandwiches.pop(0)

            after: str = "".join([str(s) for s in students])

            if before == after:
                count = len(before)
                break

        return count


def test(students: List[int], sandwiches: List[int]) -> None:
    result: int = Solution().countStudents(students.copy(), sandwiches.copy())
    print(
        f"""For these students and these sandwiches {result} students can't eat! \n -> students: {students} \n -> sandwiches: {sandwiches}""")


if __name__ == "__main__":
    test(students=[1, 1, 0, 0], sandwiches=[0, 1, 0, 1])
    test(students=[1, 1, 1, 0, 0, 1], sandwiches=[1, 0, 0, 0, 1, 1])

File: problems/1700-number-of-students-unable-to-eat-lunch/csharp/1700_number_of_students_unable_to_eat_lunch.cs
using System.Collections.Generic;
using System.Linq;
using System;

public static class Program
{
  public static void Main(string[] args)
  {
    Solution s = new Solution();
    Console.WriteLine(s.CountStudents(new int[] { 1, 1, 0, 0 }, new int[] { 0, 1, 0, 1 }));
    Console.WriteLine(s.CountStudents(new int[] { 1, 1, 1, 0, 0, 1 }, new int[] { 1, 0, 0, 0, 1, 1 }));
  }
}

public class Solution
{
  public int CountStudents(int[] studentsArr, int[] sandwichesArr)
  {
    Queue<int> students = new Queue<int>();
    foreach (int student in studentsArr)
      students.Enqueue(student);

    Stack<int> sandwiches = new Stack<int>();
    foreach (int sandwich in sandwichesArr.Reverse())
      sandwiches.Push(sandwich);

    int count = 0;
    while (students.Count != 0 && sandwiches.Count != 0)
    {
      int student = students.Peek();
      int sandwich = sandwiches.Peek();

      string before = string.Join("", students);

      if (student != sandwich)
        students.Enqueue(students.Dequeue());
      else
      {
        students.Dequeue();
        sandwiches.Pop();
      }

      string after = string.Join("", students);

      if (before == after)
      {
        count = before.Length;
        break;
      }
    }

    return count;
  }
}